### ISSUE TITLE
Old cert doesn't match new, but *only* on my dev machine

### ISSUE THREAD

**Issue by ManicMeerkat** [sparkle-project/Sparkle#612]
I spent a few days trying to figure out why I kept getting an Update Error. I kept seeing logs in the Console that look something like this:
```
9/3/15 11:31:37.086 AM PureCloud Desktop Sync[923]: Sparkle: Code signature of the new version doesn't match the old version: anchor apple generic and identifier "com.inin.purecloud.desktopsync" and (certificate leaf[field.1.2.840.113635.100.6.1.9] /* exists */ or certificate 1[field.1.2.840.113635.100.6.2.6] /* exists */ and certificate leaf[field.1.2.840.113635.100.6.1.13] /* exists */ and certificate leaf[subject.OU] = P3MPJ3YJSK). Please ensure that old and new app is signed using exactly the same certificate.


9/3/15 11:31:37.097 AM PureCloud Desktop Sync[923]: Sparkle: host info: {
    build = "1.0.37";
    format = "bundle with Mach-O thin (x86_64)";
    identifier = "com.inin.purecloud.desktopsync";
    requirements = "designated => anchor apple generic and identifier \"com.inin.purecloud.desktopsync\" and (certificate leaf[field.1.2.840.113635.100.6.1.9] /* exists */ or certificate 1[field.1.2.840.113635.100.6.2.6] /* exists */ and certificate leaf[field.1.2.840.113635.100.6.1.13] /* exists */ and certificate leaf[subject.OU] = P3MPJ3YJSK)\n";
    "signing-time" = "2015-09-02 15:27:12 +0000";
    teamid = P3MPJ3YJSK;
    version = "1.0.37 Internal (Release)";
}

9/3/15 11:31:37.098 AM PureCloud Desktop Sync[923]: Sparkle: new info: {
    build = "0.0.0";
    format = "bundle with Mach-O thin (x86_64)";
    identifier = "com.inin.purecloud.desktopsync";
    requirements = "designated => identifier \"com.inin.purecloud.desktopsync\" and anchor apple generic and certificate leaf[subject.CN] = \"Mac Developer: Jason Pollack (NTDUHXTSCP)\" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */\n";
    "signing-time" = "2015-08-27 14:47:50 +0000";
    teamid = P3MPJ3YJSK;
    version = "0.0.0 ";
}

```

I was pulling my hair out trying to figure out why the version build on the server was using the wrong certificate, which was a problem I knew I had fixed a few days ago. Then I noticed that the server version was build 0.0.0, which is clearly wrong, as I had added versioning as part of the automated build process last week. Finally today I noticed that the signing-time was last week.

"Aha!" I figured that it was pulling down either the wrong appcast, or the appcast was pointing to the wrong DMG. However, after checking carefully, both the appcast and the DMG were correct. In fact, I verified this by looking at the files in the temporary folder created by Sparkle at  ~/Library/Application Support/PureCloud Desktop Sync/.Sparkle/PureCloud Desktop Sync 1.0.41/purecloud-cmsync-internal-mac-1.0.41.dmg. The application's info.plist was showing the correct version, and after opening it in RB App Checker Lite, I verified that it had both the correct certificate and certificate signing time.

My next step was to test the update on a clean machine, and it worked perfectly. This is leading me to believe that the Sparkle updater--even though it is pulling down the correct DMG file--is comparing the certificate with a much older file located...somewhere on my machine...maybe? (Yes, I've tried rebooting, emptying my trash, searching for stray .app and .dmg files.)

Any thoughts about what is going on? I'm glad that it's working for other users, but I'd like to get my application in a state on my dev machine where the update is working again.

Thanks!
-Jason-

**Comment by zorgiepoo** [comment]
This seems kind of bizarre to me.. I notice you are using a pretty old version of Sparkle. Could you also reproduce this issue with a more recent version, perhaps?

**Comment by ManicMeerkat** [comment]
I had originally used 1.11, and rolled back to 1.8 when discovering this issue in an attempt to get things working again, since I am using 1.8 successfully in another application. 

I just tried 1.11 again, and I am getting the exact same error. The only difference is the location the files are extracted to. (I'm guessing this is how you knew it was an older version of Sparkle?)

Again, I can see the file that is being extracted, and see that it's signed correctly, and with today's date, but the log still shows the "new info" with the data from last week.

**Comment by ManicMeerkat** [comment]
I solved it. I realized I had done something very wrong, and I am seeing the side effects of it.

I discovered something very interesting. I added a couple of logs to SUCodeSigningVerifier.m:
```
    SULog(@"New bundle application path: %@", applicationPath);

    SULog(@"New bundle URL: %@", [newBundle bundleURL]);

```
On my dev machine I found this:
```
9/4/15 10:36:28.152 AM PureCloud Desktop Sync[11983]: Sparkle: New bundle application path: /Users/jasonpollack/Library/Caches/com.inin.purecloud.desktopsync/Sparkle/PureCloud Desktop Sync 1.0.43/Applications
9/4/15 10:36:28.153 AM PureCloud Desktop Sync[11983]: Sparkle: New bundle URL: file:///Applications/
```

On another machine (where the update works), I ran the same thing, and found this:

```
9/4/15 10:42:56.781 AM PureCloud Desktop Sync[8317]: Sparkle: New bundle application path: /Users/jason/Library/Caches/com.inin.purecloud.desktopsync/Sparkle/PureCloud Desktop Sync 1.0.43/PureCloud Desktop Sync.app
9/4/15 10:42:56.781 AM PureCloud Desktop Sync[8317]: Sparkle: New bundle URL: file:///Users/jason/Library/Caches/com.inin.purecloud.desktopsync/Sparkle/PureCloud%20Desktop%20Sync%201.0.43/PureCloud%20Desktop%20Sync.app/
```

Crazy, huh? 

So I added more instrumentation in Sparkle and found that it was finding the symbolic link to /Applications as a match to my bundle identifier. Clearly, this is impossible, and yet, it happened.

I recalled last week I was working on a PKG installer, and had mistakenly used /Applications as the install path rather than /Applications/PureCloud Desktop Sync.app. So what happened was that it created a Contents folder in /Applications. When the code in Sparkle tried to inspect /Applications as a bundle, it matched my bundle ID, and saw the older certificate there.

So once I found and removed the Contents folder, all started working again, and all is right with the world.

-Jason-

TL;DR I messed up a PKG installer causing Mac to think my /Applications folder was my application.
